clap_complete/src/dynamic/completer.rs

```diff
@@ -331,14 +348,69 @@ fn possible_values(a: &clap::Arg) -> Option<Vec<clap::builder::PossibleValue>> {
 ///
 /// Subcommand `rustup toolchain install` would be converted to
 /// `("install", "rustup toolchain install")`.
-fn subcommands(p: &clap::Command) -> Vec<(String, Option<StyledStr>)> {
+fn subcommands(p: &clap::Command) -> Vec<CompletionCandidate> {
     debug!("subcommands: name={}", p.get_name());
     debug!("subcommands: Has subcommands...{:?}", p.has_subcommands());
     p.get_subcommands()
         .flat_map(|sc| {
-            sc.get_name_and_visible_aliases()
-                .into_iter()
-                .map(|s| (s.to_string(), sc.get_about().cloned()))
+            sc.get_name_and_visible_aliases().into_iter().map(|s| {
+                CompletionCandidate::new(s.to_string().into())
+                    .help(sc.get_about().cloned())
+                    .visible(true)
+            })
         })
         .collect()
 }
+
+/// A completion candidate defination
+///
+/// This makes it easier to add more fields to completion candidate,
+/// rather than using `(OsString, Option<StyledStr>)` or `(String, Option<StyledStr>)` to represent a completion candidate
+#[derive(Default, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct CompletionCandidate {
+    /// Main completion candidate content
+    content: OsString,
+
+    /// Help message with a completion candidate
+    help: Option<StyledStr>,
+
+    /// Whether the completion candidate is visible
+    visible: bool,
+}
+
+impl CompletionCandidate {
+    /// Create a new completion candidate
+    pub fn new(content: OsString) -> Self {
```

Comment:
if it helps in the callers, you could do `content: impl Into<OsString>`, removing some of the `.into()` calls